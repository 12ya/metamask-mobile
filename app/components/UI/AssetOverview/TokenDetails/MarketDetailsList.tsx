import React from 'react';
import { View } from 'react-native';
import { strings } from '../../../../../locales/i18n';
import { useStyles } from '../../../../component-library/hooks';
import Text from '../../../../component-library/components/Texts/Text';
import Title from '../../../Base/Title';
import styleSheet from './TokenDetails.styles';
import { MarketDetails } from './TokenDetails';

interface MarketDetailsListProps {
  marketDetails: MarketDetails;
}

const MarketDetailsList: React.FC<MarketDetailsListProps> = ({
  marketDetails,
}) => {
  const { styles } = useStyles(styleSheet, {});

  return (
    <View style={styles.wrapper}>
      <Title style={styles.title}>{strings('token.market_details')}</Title>
      <View style={styles.listWrapper}>
        {marketDetails.marketCap && (
          <View style={[styles.listItem, styles.firstChild]}>
            <Text style={styles.listItemLabel}>
              {strings('token.market_cap')}
            </Text>
            <Text>{marketDetails.marketCap}</Text>
          </View>
        )}
        {marketDetails.totalVolume && (
          <View style={styles.listItem}>
            <Text>{strings('token.total_volume')}</Text>
            <Text>{marketDetails.totalVolume}</Text>
          </View>
        )}
        {marketDetails.volumeToMarketCap && (
          <View style={styles.listItem}>
            <Text>{strings('token.volume_to_marketcap')}</Text>
            <Text>{marketDetails.volumeToMarketCap}</Text>
          </View>
        )}
        {marketDetails.circulatingSupply && (
          <View style={styles.listItem}>
            <Text>{strings('token.circulating_supply')}</Text>
            <Text>{marketDetails.circulatingSupply}</Text>
          </View>
        )}
        {marketDetails.allTimeHigh && (
          <View style={styles.listItem}>
            <Text>{strings('token.all_time_high')}</Text>
            <Text>{marketDetails.allTimeHigh}</Text>
          </View>
        )}
        {marketDetails.allTimeLow && (
          <View style={styles.listItem}>
            <Text>{strings('token.all_time_low')}</Text>
            <Text>{marketDetails.allTimeLow}</Text>
          </View>
        )}
        {marketDetails.fullyDiluted && (
          <View style={styles.listItem}>
            <Text>{strings('token.fully_diluted')}</Text>
            <Text>{marketDetails.fullyDiluted}</Text>
          </View>
        )}
      </View>
    </View>
  );
};

export default MarketDetailsList;
