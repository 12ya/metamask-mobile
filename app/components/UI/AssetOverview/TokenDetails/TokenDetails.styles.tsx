import type { Theme } from '@metamask/design-tokens';
import { StyleSheet, TextStyle } from 'react-native';

const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;
  const { typography, colors } = theme;
  return StyleSheet.create({
    wrapper: {
      marginTop: 20,
    },
    contentWrapper: {
      paddingVertical: 4,
    },
    text: {
      ...typography.sBodyMD,
      marginVertical: 0,
    } as TextStyle,
    title: {
      ...typography.sHeadingSM,
      marginVertical: 0,
      //   marginBottom: 4,
      paddingVertical: 8,
    } as TextStyle,
    icon: { marginLeft: 4 },
    listWrapper: {
      paddingTop: 8,
      paddingBottom: 8,
    },
    listItem: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      paddingVertical: 4,
    },
    listItemStacked: {
      flexDirection: 'column',
      justifyContent: 'space-between',
      paddingVertical: 4,
    },
    firstChild: {
      paddingTop: 0,
    },
    lastChild: {
      paddingBottom: 0,
    },
    listItemLabel: {
      ...typography.sBodyMDMedium,
      marginVertical: 0,
    } as TextStyle,
    listItemValue: {
      ...typography.sBodyMD,
      marginVertical: 0,
    } as TextStyle,
    copyButton: {
      flexDirection: 'row',
      alignItems: 'center',
      backgroundColor: colors.primary.muted,
      borderRadius: 20,
      paddingHorizontal: 8,
      marginLeft: 8,
    },
  });
};

export default styleSheet;
