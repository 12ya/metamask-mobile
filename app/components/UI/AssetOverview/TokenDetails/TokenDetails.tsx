import { zeroAddress } from 'ethereumjs-util';
import React from 'react';
import { TouchableOpacity, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { showAlert } from '../../../../actions/alert';
import i18n, { strings } from '../../../../../locales/i18n';
import { useStyles } from '../../../../component-library/hooks';
import Text, {
  TextColor,
  TextVariant,
} from '../../../../component-library/components/Texts/Text';
import Title from '../../../Base/Title';
import { TokenDescriptions } from '../../../hooks/useTokenDescriptions';
import { Asset } from '../AssetOverview.types';
import styleSheet from './TokenDetails.styles';
import Icon, {
  IconColor,
  IconName,
  IconSize,
} from '../../../../component-library/components/Icons/Icon';
import { formatAddress, safeToChecksumAddress } from '../../../../util/address';
import ClipboardManager from '../../../../core/ClipboardManager';
import { selectTokenList } from '../../../../selectors/tokenListController';
import { selectContractExchangeRates } from '../../../../selectors/tokenRatesController';
import {
  selectConversionRate,
  selectCurrentCurrency,
} from '../../../../selectors/currencyRateController';
import {
  convertDecimalToPercentage,
  localizeLargeNumber,
} from '../../../../util/number';
import { formatCurrency } from '../../../../util/confirm-tx';

interface TokenDetailsProps {
  asset: Asset;
}

interface TokenDetailsListProps {
  tokenDetails: TokenDetails;
}

interface TokenDetails {
  contractAddress: string | null;
  tokenDecimal: number | null;
  tokenList: string | null;
}

interface MarketDetailsListProps {
  marketDetails: MarketDetails;
}

interface MarketDetails {
  marketCap: string | null;
  totalVolume: number | null;
  volumeToMarketCap: string | null;
  circulatingSupply: number | null;
  allTimeHigh: string | null;
  allTimeLow: string | null;
  fullyDiluted: number | null;
}

const skeletonProps = {
  width: '100%',
  height: 18,
  borderRadius: 6,
  marginBottom: 8,
};

const TokenDetails: React.FC<TokenDetailsProps> = ({ asset }) => {
  const { styles } = useStyles(styleSheet, {});
  const tokenList = useSelector(selectTokenList);
  const tokenExchangeRates = useSelector(selectContractExchangeRates);
  const conversionRate = useSelector(selectConversionRate);
  const currentCurrency = useSelector(selectCurrentCurrency);

  const tokenContractAddress = safeToChecksumAddress(asset.address);

  if (!tokenContractAddress) {
    console.log("can't find contract address");
    return null;
  }

  if (!conversionRate || conversionRate < 0) {
    console.log('invalid conversion rate');
    return null;
  }

  const tokenMetadata = tokenList[tokenContractAddress.toLowerCase()];
  const marketData = tokenExchangeRates[tokenContractAddress];

  // why isn't zeroAddress() in TokenList?
  const tokenDetails: TokenDetails = asset.isETH
    ? {
        contractAddress: formatAddress(zeroAddress(), 'short'),
        tokenDecimal: 18,
        tokenList: '',
      }
    : {
        contractAddress: formatAddress(tokenContractAddress, 'short') || null,
        tokenDecimal: tokenMetadata.decimals || null,
        tokenList: tokenMetadata.aggregators.join(', ') || null,
      };

  const marketDetails: MarketDetails = {
    marketCap:
      marketData?.marketCap > 0
        ? localizeLargeNumber(i18n, conversionRate * marketData.marketCap)
        : null,
    totalVolume:
      marketData?.totalVolume > 0
        ? localizeLargeNumber(i18n, conversionRate * marketData.totalVolume)
        : null,
    volumeToMarketCap:
      marketData?.marketCap > 0
        ? convertDecimalToPercentage(
            marketData.totalVolume / marketData.marketCap,
          )
        : null,
    circulatingSupply:
      marketData?.circulatingSupply > 0
        ? localizeLargeNumber(i18n, marketData.circulatingSupply)
        : null,
    allTimeHigh:
      marketData?.allTimeHigh > 0
        ? formatCurrency(
            conversionRate * marketData.allTimeHigh,
            currentCurrency,
          )
        : null,
    allTimeLow:
      marketData?.allTimeLow > 0
        ? formatCurrency(
            conversionRate * marketData.allTimeLow,
            currentCurrency,
          )
        : null,
    fullyDiluted:
      marketData?.dilutedMarketCap > 0
        ? localizeLargeNumber(i18n, marketData.dilutedMarketCap)
        : null,
  };

  return (
    <View style={styles.wrapper}>
      {(asset.isETH || tokenMetadata) && (
        <TokenDetailsList tokenDetails={tokenDetails} />
      )}
      {marketData && <MarketDetailsList marketDetails={marketDetails} />}
    </View>
  );
};

const TokenDetailsList: React.FC<TokenDetailsListProps> = ({
  tokenDetails,
}) => {
  const { styles } = useStyles(styleSheet, {});
  const dispatch = useDispatch();

  const handleShowAlert = (config: {
    isVisible: boolean;
    autodismiss: number;
    content: string;
    data: { msg: string };
  }) => dispatch(showAlert(config));

  const copyAccountToClipboard = async () => {
    await ClipboardManager.setString(tokenDetails.contractAddress);

    handleShowAlert({
      isVisible: true,
      autodismiss: 1500,
      content: 'clipboard-alert',
      data: { msg: strings('account_details.account_copied_to_clipboard') },
    });
  };

  return (
    <View style={styles.wrapper}>
      <Title style={styles.title}>{strings('token.token_details')}</Title>
      <View style={styles.listWrapper}>
        {tokenDetails.contractAddress && (
          <View style={[styles.listItem, styles.firstChild]}>
            <Text style={styles.listItemLabel}>
              {strings('token.contract_address')}
            </Text>
            <TouchableOpacity
              style={styles.copyButton}
              onPress={copyAccountToClipboard}
            >
              <Text color={TextColor.Primary} variant={TextVariant.BodySM}>
                {tokenDetails.contractAddress}
              </Text>
              <Icon
                name={IconName.Copy}
                size={IconSize.Sm}
                color={IconColor.Primary}
                style={styles.icon}
              />
            </TouchableOpacity>
          </View>
        )}
        {tokenDetails.tokenDecimal && (
          <View style={styles.listItem}>
            <Text>{strings('token.token_decimal')}</Text>
            <Text>{tokenDetails.tokenDecimal}</Text>
          </View>
        )}
        {tokenDetails.tokenList && (
          <View style={[styles.listItemStacked, styles.lastChild]}>
            <Text>{strings('token.token_list')}</Text>
            <Text>{tokenDetails.tokenList}</Text>
          </View>
        )}
      </View>
    </View>
  );
};

const MarketDetailsList: React.FC<MarketDetailsListProps> = ({
  marketDetails,
}) => {
  const { styles } = useStyles(styleSheet, {});

  return (
    <View style={styles.wrapper}>
      <Title style={styles.title}>Market Details</Title>
      <View style={styles.listWrapper}>
        {marketDetails.marketCap && (
          <View style={[styles.listItem, styles.firstChild]}>
            <Text style={styles.listItemLabel}>
              {strings('token.market_cap')}
            </Text>
            <Text>{marketDetails.marketCap}</Text>
          </View>
        )}
        {marketDetails.totalVolume && (
          <View style={styles.listItem}>
            <Text>{strings('token.total_volume')}</Text>
            <Text>{marketDetails.totalVolume}</Text>
          </View>
        )}
        {marketDetails.volumeToMarketCap && (
          <View style={styles.listItem}>
            <Text>{strings('token.volume_to_marketcap')}</Text>
            <Text>{marketDetails.volumeToMarketCap}</Text>
          </View>
        )}
        {marketDetails.circulatingSupply && (
          <View style={styles.listItem}>
            <Text>{strings('token.circulating_supply')}</Text>
            <Text>{marketDetails.circulatingSupply}</Text>
          </View>
        )}
        {marketDetails.allTimeHigh && (
          <View style={styles.listItem}>
            <Text>{strings('token.all_time_high')}</Text>
            <Text>{marketDetails.allTimeHigh}</Text>
          </View>
        )}
        {marketDetails.allTimeLow && (
          <View style={styles.listItem}>
            <Text>{strings('token.all_time_low')}</Text>
            <Text>{marketDetails.allTimeLow}</Text>
          </View>
        )}
        {marketDetails.fullyDiluted && (
          <View style={styles.listItem}>
            <Text>{strings('token.fully_diluted')}</Text>
            <Text>{marketDetails.fullyDiluted}</Text>
          </View>
        )}
      </View>
    </View>
  );
};

export default TokenDetails;
