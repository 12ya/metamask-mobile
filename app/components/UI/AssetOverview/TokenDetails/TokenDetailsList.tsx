import React from 'react';
import { TouchableOpacity, View } from 'react-native';
import { useDispatch } from 'react-redux';
import { showAlert } from '../../../../actions/alert';
import { strings } from '../../../../../locales/i18n';
import { useStyles } from '../../../../component-library/hooks';
import Text, {
  TextColor,
  TextVariant,
} from '../../../../component-library/components/Texts/Text';
import Title from '../../../Base/Title';
import styleSheet from './TokenDetails.styles';
import Icon, {
  IconColor,
  IconName,
  IconSize,
} from '../../../../component-library/components/Icons/Icon';
import ClipboardManager from '../../../../core/ClipboardManager';
import { TokenDetails } from './TokenDetails';

interface TokenDetailsListProps {
  tokenDetails: TokenDetails;
}

const TokenDetailsList: React.FC<TokenDetailsListProps> = ({
  tokenDetails,
}) => {
  const { styles } = useStyles(styleSheet, {});
  const dispatch = useDispatch();

  const handleShowAlert = (config: {
    isVisible: boolean;
    autodismiss: number;
    content: string;
    data: { msg: string };
  }) => dispatch(showAlert(config));

  const copyAccountToClipboard = async () => {
    await ClipboardManager.setString(tokenDetails.contractAddress);

    handleShowAlert({
      isVisible: true,
      autodismiss: 1500,
      content: 'clipboard-alert',
      data: { msg: strings('account_details.account_copied_to_clipboard') },
    });
  };

  return (
    <View style={styles.wrapper}>
      <Title style={styles.title}>{strings('token.token_details')}</Title>
      <View style={styles.listWrapper}>
        {tokenDetails.contractAddress && (
          <View style={[styles.listItem, styles.firstChild]}>
            <Text style={styles.listItemLabel}>
              {strings('token.contract_address')}
            </Text>
            <TouchableOpacity
              style={styles.copyButton}
              onPress={copyAccountToClipboard}
            >
              <Text color={TextColor.Primary} variant={TextVariant.BodySM}>
                {tokenDetails.contractAddress}
              </Text>
              <Icon
                name={IconName.Copy}
                size={IconSize.Sm}
                color={IconColor.Primary}
                style={styles.icon}
              />
            </TouchableOpacity>
          </View>
        )}
        {tokenDetails.tokenDecimal && (
          <View style={styles.listItem}>
            <Text>{strings('token.token_decimal')}</Text>
            <Text>{tokenDetails.tokenDecimal}</Text>
          </View>
        )}
        {tokenDetails.tokenList && (
          <View style={[styles.listItemStacked, styles.lastChild]}>
            <Text>{strings('token.token_list')}</Text>
            <Text>{tokenDetails.tokenList}</Text>
          </View>
        )}
      </View>
    </View>
  );
};

export default TokenDetailsList;
