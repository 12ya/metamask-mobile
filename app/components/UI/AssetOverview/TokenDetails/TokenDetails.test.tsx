import React from 'react';
import renderWithProvider from '../../../../util/test/renderWithProvider';
import { backgroundState } from '../../../../util/test/initial-root-state';
import Engine from '../../../../core/Engine';
import TokenDetails from './';
const mockedEngine = Engine;

const mockAsset = {
  address: '0x6B175474E89094C44Da98b954EedeAC495271d0F',
  aggregators: [
    'Metamask',
    'Aave',
    'Bancor',
    'CMC',
    'Crypto.com',
    'CoinGecko',
    '1inch',
    'PMM',
    'Sushiswap',
    'Zerion',
    'Lifi',
    'Socket',
    'Squid',
    'Openswap',
    'Sonarwatch',
    'UniswapLabs',
    'Coinmarketcap',
  ],
  balanceError: undefined,
  balanceFiat: '$6.49',
  decimals: 18,
  image:
    'https://static.cx.metamask.io/api/v1/tokenIcons/1/0x6b175474e89094c44da98b954eedeac495271d0f.png',
  name: 'Dai Stablecoin',
  symbol: 'DAI',
  isETH: false,
};

jest.mock('../../../../core/Engine.ts', () => ({
  init: () => mockedEngine.init({}),
  context: {
    KeyringController: {
      getQRKeyringState: async () => ({ subscribe: () => ({}) }),
    },
    TokenListController: {
      tokenList: {
        '0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f': {
          address: '0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f',
          symbol: 'SNX',
          decimals: 18,
          name: 'Synthetix Network Token',
          iconUrl:
            'https://static.cx.metamask.io/api/v1/tokenIcons/1/0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f.png',
          type: 'erc20',
          aggregators: [
            'Aave',
            'Bancor',
            'CMC',
            'Crypto.com',
            'CoinGecko',
            '1inch',
            'PMM',
            'Synthetix',
            'Zerion',
            'Lifi',
          ],
          occurrences: 10,
          fees: {
            '0x5fd79d46eba7f351fe49bff9e87cdea6c821ef9f': 0,
            '0xda4ef8520b1a57d7d63f1e249606d1a459698876': 0,
          },
        },
      },
      tokensChainsCache: {},
      preventPollingOnNetworkRestart: false,
    },
  },
}));

const initialState = {
  engine: {
    backgroundState,
  },
};

describe('TokenDetails', () => {
  it('should render correctly', () => {
    const { toJSON } = renderWithProvider(<TokenDetails asset={mockAsset} />, {
      state: initialState,
    });
    console.log(toJSON());
    expect(toJSON()).toMatchSnapshot();
  });
});
