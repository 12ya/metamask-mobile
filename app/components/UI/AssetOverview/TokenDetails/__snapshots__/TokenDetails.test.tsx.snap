// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenDetails should render correctly 1`] = `null`;
